provider/postgis/postgis.go

```diff
@@ -725,7 +740,23 @@ func (p Provider) inspectLayerGeomType(l *Layer) error {
 		return err
 	}
 
-	rows, err := p.pool.Query(context.Background(), sql)
+	// substitute default values to parameter
+	params := extractQueryParamValues(pname, maps, l)
+	if err != nil {
+		return err
+	}
+
+	args := make([]interface{}, 0)
+	sql = provider.ReplaceParams(params, sql, &args)
+
+	if provider.ParameterTokenRegexp.MatchString(sql) {
+		// remove all parameter tokens for inspection
+		// crossing our fingers that the query is still valid 🤞
+		// if not, the user will have to specify `geometry_type` in the config
```

Comment:
> if `default_sql` is defined, use it (it seems that this code doesn't do that)

It does, `*QueryParameter.ToDefaultValue` handles this.

> if neither is defined and the parameter is optional/required

If neither is defined, the parameter is required as long as its token occurs in an SQL expression. We don't have an explicit `required` parameter now, and I don't think we should. The current approach have only one invalid invariant: when you specify both `defaul_value` and `defaul_sql`. Having the `required` config option will produce much more without giving any benefits IMO.

Anyway, this does behave the same way as during request processing (we use the same `*QueryParameter.ToDefaultValue` method), except that it tries to go an extra step further: tries to remove tokens that have left after resolving built-in one and custom with defaults. This could work in some cases that I consider pretty common, e.g.,

```
SELECT … WHERE wkb_geometry && !BBOX! !PARAM!
```
```
  [[maps.params]]
  name    = "param"
  token   = "!PARAM!"
  type = "string"
  sql = "AND osm_id == ?"
```